bin/runtest_common.ml

```diff
@@ -4,103 +4,221 @@ module Test_kind = struct
   ;;
 end
 
-let cram_tests_of_dir parent_dir =
+(** [classify_ml_test ~sctx ~dir ~ml_file] looks up whether [ml_file] is part
+    of a (tests) or (library (inline_tests ...)) stanza in [dir] and returns:
+
+    - [Ok (`Test_executable exe_name)] if the file is a test entry point, or if
+      the file belongs to a tests stanza with a single entry point (in which
+      case that entry point is returned)
+
+    - [Ok (`Inline_tests_library lib_name)] if the file belongs to a library
+      with inline tests enabled
+
+    - [Error `Not_an_entry_point] if the file belongs to a tests stanza with
+      multiple entry points but is not itself an entry point
+
+    - [Error `Not_a_test] if the file is not part of any tests stanza or inline
+      tests library *)
+let classify_ml_test ~sctx ~dir ~ml_file =
   let open Memo.O in
-  Source_tree.find_dir parent_dir
-  >>= function
-  | None -> Memo.return []
-  | Some dir -> Dune_rules.Cram_rules.cram_tests dir
+  let ml_file_no_ext = Filename.remove_extension ml_file in
+  match Dune_lang.Module_name.of_string_opt ml_file_no_ext with
+  | None -> Memo.return (Error `Not_a_test)
+  | Some module_name ->
+    let* dir_contents =
+      Dir_contents.get
+        sctx
+        ~dir:
+          (Path.Build.append_source (Context.build_dir (Super_context.context sctx)) dir)
+    in
+    let* ml_sources = Dir_contents.ocaml dir_contents
+    and* scope = Dir_contents.dir dir_contents |> Dune_rules.Scope.DB.find_by_dir in
+    Dune_rules.Ml_sources.find_origin
+      ml_sources
+      ~libs:(Dune_rules.Scope.libs scope)
+      [ module_name ]
+    >>| (function
+     | Some (Library lib) when Library.has_inline_tests lib ->
+       let lib_name = snd lib.name |> Lib_name.Local.to_string in
+       Ok (`Inline_tests_library lib_name)
+     | Some (Library _ | Executables _ | Melange _) | None -> Error `Not_a_test
+     | Some (Tests { exes; _ }) ->
+       let exe_names = Nonempty_list.to_list exes.names |> List.map ~f:snd in
+       if List.mem exe_names ml_file_no_ext ~equal:String.equal
+       then Ok (`Test_executable ml_file_no_ext)
+       else (
+         match exe_names with
+         | [ single_exe ] -> Ok (`Test_executable single_exe)
+         | [] | _ :: _ :: _ -> Error `Not_an_entry_point))
 ;;
 
-let find_cram_test cram_tests path =
-  List.find_map cram_tests ~f:(function
-    | Ok cram_test when Path.Source.equal path (Source.Cram_test.path cram_test) ->
-      Some cram_test
-    (* We raise any error we encounter when looking for our test specifically. *)
-    | Error (Dune_rules.Cram_rules.Missing_run_t cram_test)
-      when Path.Source.equal path (Source.Cram_test.path cram_test) ->
-      Dune_rules.Cram_rules.missing_run_t cram_test
+(** Looks up a cram test at [path] in [parent_dir]. Returns [Some Test_kind.t]
+    if found, [None] otherwise. Raises if the test exists but is missing its
+    run.t file. *)
+let find_cram ~parent_dir path =
```

Comment:
Why do we need this change?

Reply:
Before we had two functions. One for getting all the cram tests in a directory and another for picking out a specific cram test from a set of them. This function is basically both stuck together. It fits more inline with what the other `find_ml` test is doing.